packages/firestore/src/core/view.ts

```diff
@@ -153,53 +205,72 @@ export class View {
           );
           newDoc = this.query.matches(newDoc) ? newDoc : null;
         }
-        if (newDoc) {
-          newDocumentSet = newDocumentSet.add(newDoc);
-          if (newDoc.hasLocalMutations) {
-            newMutatedKeys = newMutatedKeys.add(key);
-          } else {
-            newMutatedKeys = newMutatedKeys.delete(key);
-          }
-        } else {
-          newDocumentSet = newDocumentSet.delete(key);
-          newMutatedKeys = newMutatedKeys.delete(key);
-        }
+
+        const oldDocHadPendingMutations = oldDoc
+          ? this.mutatedKeys.has(oldDoc.key)
+          : false;
+        const newDocHasPendingMutations = newDoc
+          ? newDoc.hasLocalMutations ||
+            // We only consider committed mutations for documents that were
+            // mutated during the lifetime of the view.
+            (this.mutatedKeys.has(newDoc.key) && newDoc.hasCommittedMutations)
+          : false;
+
+        let changeApplied = false;
 
         // Calculate change
         if (oldDoc && newDoc) {
           const docsEqual = oldDoc.data.isEqual(newDoc.data);
-          if (
-            !docsEqual ||
-            oldDoc.hasLocalMutations !== newDoc.hasLocalMutations
-          ) {
-            // only report a change if document actually changed
-            if (docsEqual) {
-              changeSet.track({ type: ChangeType.Metadata, doc: newDoc });
-            } else {
-              changeSet.track({ type: ChangeType.Modified, doc: newDoc });
-            }
+          if (!docsEqual) {
+            if (!this.shouldWaitForRemoteVersion(oldDoc, newDoc)) {
+              changeSet.track({
+                type: ChangeType.Modified,
+                doc: newDoc
+              });
+              changeApplied = true;
 
-            if (
-              lastDocInLimit &&
-              this.query.docComparator(newDoc, lastDocInLimit) > 0
-            ) {
-              // This doc moved from inside the limit to after the limit.
-              // That means there may be some doc in the local cache that's
-              // actually less than this one.
-              needsRefill = true;
+              if (
```

Comment:
It's simpler now to do this check only in this branch (since we otherwise also have to include `changeApplied` in the if-check). I can't think of a case where a document would fall outside the limit based on a metadata-only change.